### PR TITLE
[SYCL] Emit integration header even if no kernels provided

### PR DESCRIPTION
**Pull request by Fznamznon**
Since device_global variables can be declared in one file and used in
another, it is possible to receive an input with device_global and no
kernels. Make sure that integration header and footer are emitted in this
case.

### CODE REVIEW

File: clang/test/CodeGenSYCL/int_header_without_kernels.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// RUN: %clang_cc1 -fsycl-is-device -std=c++17 -internal-isystem %S/Inputs -triple spir64-unknown-unknown -fsycl-int-footer=%t.footer.h -fsycl-int-header=%t.header.h -fsycl-unique-prefix=THE_PREFIX %s -emit-llvm -o %t.ll
```
Comment (premanandrao) [inline clang/test/CodeGenSYCL/int_header_without_kernels.cpp:1]:
```suggestion
// RUN: %clang_cc1 -fsycl-is-device -std=c++17 -internal-isystem %S/Inputs -triple spir64-unknown-unknown -fsycl-int-footer=%t.footer.h -fsycl-int-header=%t.header.h %s -emit-llvm -o %t.ll
```

**Review by premanandrao — COMMENTED** [review]
LGTM, just a minor nit.

File: clang/test/CodeGenSYCL/int_header_without_kernels.cpp (hunk 1 above)
Comment (Fznamznon) [reply clang/test/CodeGenSYCL/int_header_without_kernels.cpp]: Good catch, thank you!

File: clang/include/clang/Sema/Sema.h [hunk 2]
```diff
@@ -433,21 +435,21 @@ class SYCLIntegrationHeader {
   llvm::SmallVector<SpecConstID, 4> SpecConsts;
 
   Sema &S;
+
+  bool NeedToEmitDeviceGlobalMap = false;
```
Comment (smanna12) [inline clang/include/clang/Sema/Sema.h:439]: Same here

File: clang/include/clang/Sema/Sema.h [hunk 3]
```diff
@@ -359,6 +359,8 @@ class SYCLIntegrationHeader {
     Itr->updateKernelNames(Name, StableName);
   }
 
+  void addDeviceGlobalMap() { NeedToEmitDeviceGlobalMap = true; }
```
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/include/clang/Sema/Sema.h]: Sure, done. I've added the comments. Making these comments made me re-think and rename method and field member names as well.

**Review by smanna12 — APPROVED** [review]
LGTM

File: clang/include/clang/Sema/Sema.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: Thank you. I like rename better now.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -4656,7 +4656,7 @@ void SYCLIntegrationHeader::emit(raw_ostream &O) {
   // whose sole purpose is to run its constructor before the application's
   // main() function.
 
-  if (S.getSyclIntegrationFooter().isDeviceGlobalsEmitted()) {
+  if (NeedToEmitDeviceGlobalRegistration) {
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:4659]: Can you explain why this change forces emission? Prior to this PR, the variable signaling emit required was maintained in footer and now it is in header. How did changing this solve the issue?

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:4659]: Integration header is created when it is accessed via `Sema::getSyclIntegrationHeader()`. Prior to this change this access happened only when info about kernels was communicated to the integration header, so the integration header was not emitted if there was no kernels in user's code. This change forces creation of the header by calling of `Sema::getSyclIntegrationHeader()` by the function that emits integration footer (see line 5108 in SemaSYCL.cpp in this diff). Moving of variable signaling that `__sycl_device_global_registration` is required from footer to header seemed logical choice here, so now the footer communicates some info to the header and therefore forces emission of the header.

Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:4659]: Thanks for explaining!

**Review by elizabethandrews — APPROVED** [review]

**Comment by Fznamznon** [comment]
Thanks for reviews everyone!

**Comment by Fznamznon** [comment]
@intel/llvm-gatekeepers , I believe this one is ready for merge.

The following XPASS on GPU:
Unexpectedly Passed Tests (1):
  SYCL :: KernelAndProgram/undefined-symbol.cpp
Is not related to the patch and it seems it is being fixed by https://github.com/intel/llvm-test-suite/pull/958/files .
